### DISCUSSION TITLE
다음 JavaScript 코드를 보고, 오류를 찾고 수정해 보세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#105] [모듈]
- 다음 JavaScript 코드를 보고, 오류를 찾고 수정해 보세요.
  ```jsx
  // math.js
  export function add(a, b) {
    return a + b;
  }
  
  export function subtract(a, b) {
    return a - b;
  }
  
  // main.js
  import add from './math.js';
  
  console.log(add(2, 3));
  ```
  - default export와 named export를 함께 사용할 수 있나요?
  - default export는 여러개 사용 가능한가요?
  - 브라우저 환경에서 동작할 때 html 파일이 아래와 같아도 위의 오류를 수정한 코드가 정상 동작할까요?
    ```jsx
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>브라우저 모듈 예제</title>
    </head>
    <body>
      <script src="main.js"></script>
    </body>
    </html>
    ```
    - 정상 동작하려면 어떻게 처리해줘야 할까요?
    - math.js 파일도 모듈인데 script 태그로 안 불러와도 되나요?
  - node.js 환경이라 가정했을 때도 위의 오류를 수정한 코드의 실행이 정상적으로 동작할까요?
    - 정상적으로 동작하지 않는다면 어떻게 수정해야 할까요?
    - 아래 코드의 경우는 정상 동작할까요? 이유는 뭔가요?
       ```jsx
        // math.js
        module.exports = {
          add: (a, b) => a + b,
          subtract: (a, b) => a - b,
        };
        
        // main.js
        const math = require("./math");
        console.log(math.add(2, 3));
        ```

**Comment by publdaze** [comment]
> * 다음 JavaScript 코드를 보고, 오류를 찾고 수정해 보세요.
>   ```js
>   // math.js
>   export function add(a, b) {
>     return a + b;
>   }
>   
>   export function subtract(a, b) {
>     return a - b;
>   }
>   
>   // main.js
>   import add from './math.js';
>   
>   console.log(add(2, 3));
>   ```
- name export를 사용하면 import할 때 중괄호를 사용해야 함 -> main.js의 import 문을 `import { add } from './math.js';`
- 중괄호 없이 그대로 사용하려면 export 방식을 default export로 바꾸면 됨 -> math.js의 add를 `default export function add(a, b)`
>   * default export와 named export를 함께 사용할 수 있나요?
- 가능, 하나의 모듈에서 주요 기능은 default export로, 보조 기능은 named export로 내보낼 수 있고, 모듈을 가져올 때는 default export는 별칭으로, named export는 중괄호로 감싸서 가져올 수 있음
>   * default export는 여러개 사용 가능한가요?
- 불가능, 한 모듈에는 하나의 default export만 가능
>   * 브라우저 환경에서 동작할 때 html 파일이 아래와 같아도 위의 오류를 수정한 코드가 정상 동작할까요?
>
>     ```js
>     <!DOCTYPE html>
>     <html lang="en">
>     <head>
>       <meta charset="UTF-8">
>       <meta name="viewport" content="width=device-width, initial-scale=1.0">
>       <title>브라우저 모듈 예제</title>
>     </head>
>     <body>
>       <script src="main.js"></script>
>     </body>
>     </html>
>     ```
>
>  - 정상 동작하려면 어떻게 처리해줘야 할까요?
- script 태그에서 type="module" 어트리뷰트가 없으므로 정상 동작하지 않음, 해당 어트리뷰트가 추가되어가 모듈로써 정상동작 가능
>  - math.js 파일도 모듈인데 script 태그로 안 불러와도 되나요?
- math.js는 import되는 의존성이므로 최상위인 main만 불러와도 모듈로 동작 가능
>  - node.js 환경이라 가정했을 때도 위의 오류를 수정한 코드의 실행이 정상적으로 동작할까요?
- node.js 환경에서는 commonjs를 사용하므로 es6 모듈 방식인 위의 코드는 정상동작하지 않음
>  - 정상적으로 동작하지 않는다면 어떻게 수정해야 할까요?
- 파일 이름을 mjs로 변경하거나, package.json 파일에 { type: module }을 추가하면 es6 모듈 방식 사용 가능
>  - 아래 코드의 경우는 정상 동작할까요? 이유는 뭔가요?
>
>       ```js
>        // math.js
>        module.exports = {
>          add: (a, b) => a + b,
>          subtract: (a, b) => a - b,
>        };
>        
>        // main.js
>        const math = require("./math");
>        console.log(math.add(2, 3));
>       ```
>
- commonjs 모듈에 해당하는 구문이므로, commonjs 모듈을 사용하는 node.js 환경에서 정상 동작

**Comment by pipisebastian** [comment]
> * 다음 JavaScript 코드를 보고, 오류를 찾고 수정해 보세요.
>   ```js
>   // math.js
>   export function add(a, b) {
>     return a + b;
>   }
>   
>   export function subtract(a, b) {
>     return a - b;
>   }
>   
>   // main.js
>   import add from './math.js';
>   
>   console.log(add(2, 3));
>   ```


- named export 에서는 import할때 {} 중괄호를 사용해야합니다.
- 중괄호를 사용하거나, 혹은 default export를 사용하여, 내보내면 됩니다.   


>   * default export와 named export를 함께 사용할 수 있나요?
- 가능합니다
```js
import defaultImport, { namedExport1, namedExport2 } from './module.js';

```
>   * default export는 여러개 사용 가능한가요?

- 불가능합니다. default export는 하나의 값만 내보낼 수 있습니다.

>   * 브라우저 환경에서 동작할 때 html 파일이 아래와 같아도 위의 오류를 수정한 코드가 정상 동작할까요?
>
>     ```js
>     <!DOCTYPE html>
>     <html lang="en">
>     <head>
>       <meta charset="UTF-8">
>       <meta name="viewport" content="width=device-width, initial-scale=1.0">
>       <title>브라우저 모듈 예제</title>
>     </head>
>     <body>
>       <script src="main.js"></script>
>     </body>
>     </html>
>     ```     

- 불가능합니다.
  
>   * 정상 동작하려면 어떻게 처리해줘야 할까요?

- script 태그에 type = “module” 어트리뷰트 추가하면 됩니다.
- 
>   * math.js 파일도 모듈인데 script 태그로 안 불러와도 되나요?

- main 내부에서 math 모듈을 불러오므로, main만 불러오면 됩니다.

>   * node.js 환경이라 가정했을 때도 위의 오류를 수정한 코드의 실행이 정상적으로 동작할까요?

- 불가능합니다. node.js 환경은 기본적으로 CommonJs방식을 사용합니다.

>   * 정상적으로 동작하지 않는다면 어떻게 수정해야 할까요?

- package.json 설정: type을 module로 설정하는 방법.
- 파일 확장자를 '.mjs'로 사용
 
>   * 아래 코드의 경우는 정상 동작할까요? 이유는 뭔가요?
>
>       ```js
>        // math.js
>        module.exports = {
>          add: (a, b) => a + b,
>          subtract: (a, b) => a - b,
>        };
>        
>        // main.js
>        const math = require("./math");
>        console.log(math.add(2, 3));
>       ```

- CommonJs 방식으로, node.js환경에서는 작동합니다.
